internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -296,10 +301,18 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 		// use the new agent executable as it should be able to parse the new update marker
 		watcherExecutable = paths.BinaryPath(filepath.Join(paths.Top(), unpackRes.VersionedHome), agentName)
 	}
-	if err := InvokeWatcher(u.log, watcherExecutable); err != nil {
+	var watcherCmd *exec.Cmd
+	if watcherCmd, err = InvokeWatcher(u.log, watcherExecutable); err != nil {
 		u.log.Errorw("Rolling back: starting watcher failed", "error.message", err)
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
-		return nil, err
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(err, rollbackErr)
+	}
+
+	watcherWaitErr := waitForWatcher(ctx, u.log, markerFilePath(paths.Data()), 30*time.Second)
+	if watcherWaitErr != nil {
+		killWatcherErr := watcherCmd.Process.Kill()
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(watcherWaitErr, killWatcherErr, rollbackErr)
```

Comment:
This chunk of code seems like an implementation detail of `InvokeWatcher`.  WDYT about calling it from within `InvokeWatcher` instead of having `InvokeWatcher` return `watcherCmd`?

The only reason I can think of not doing it the way I'm suggesting is if we don't want the other call to `InvokeWatcher` (when Agent starts up) to not do the same waiting logic that you have in this chunk.

Reply:
Thanks, that makes sense.

We could still encapsulate this implementation into `InvokeWatcher` and have it take a boolean flag on whether it should wait for the watcher to start or not, but I feel less strongly now about doing it that way vs. the way it's done here.

Reply:
Single responsibility principle made me choose having a function for waiting separate from the rest. It should be also easier to test like this 😉